step-automation-packages/step-automation-packages-manage/src/main/java/step/automation/packages/AutomationPackageManager.java

```diff
@@ -0,0 +1,372 @@
+        List<Function> completeFunctions = prepareFunctionsStaging(newPackage, automationPackageArchive, packageContent, enricher, oldPackage);
+
+        // delete old package entities
+        deleteAutomationPackageEntities(oldPackage);
+
+        // persist all staged entities
+        persistStagedEntities(completeExecTasksParameters, completeFunctions, completePlans);
+
+        // save automation package metadata
+        String result =  automationPackageAccessor.save(newPackage).getId().toString();
+
+        log.info("Automation package has been updated ({}). Plans: {}. Functions: {}. Schedules: {}", newPackage.getAttribute(AbstractOrganizableObject.NAME), completePlans.size(), completeFunctions.size(), completeExecTasksParameters.size());
+        return result;
+    }
+
+    private void persistStagedEntities(List<ExecutiontTaskParameters> completeExecTasksParameters, List<Function> completeFunctions, List<Plan> completePlans) throws SetupFunctionException, FunctionTypeException {
+        for (Function completeFunction : completeFunctions) {
+            functionManager.saveFunction(completeFunction);
+        }
+
+        for (Plan plan : completePlans) {
+            planAccessor.save(plan);
+        }
+
+        for (ExecutiontTaskParameters execTasksParameter : completeExecTasksParameters) {
+            executionScheduler.addExecutionTask(execTasksParameter, false);
+        }
+
+    }
+
+    protected List<Plan> preparePlansStaging(AutomationPackage newPackage, AutomationPackageContent packageContent, AutomationPackage oldPackage, ObjectEnricher enricher) {
+        List<Plan> oldPlans = oldPackage != null ? getPackagePlans(oldPackage) : new ArrayList<>();
+        Map<String, ObjectId> planNameToIdMap = createNameToIdMap(oldPlans);
+
+        List<Plan> completePlans = new ArrayList<>();
+        for (Plan plan : packageContent.getPlans()) {
+            // keep old id
+            ObjectId oldId = planNameToIdMap.get(plan.getAttribute(AbstractOrganizableObject.NAME));
+            if (oldId != null) {
+                plan.setId(oldId);
+            }
+
+            plan.addCustomField(AutomationPackageEntity.AUTOMATION_PACKAGE_ID, newPackage.getId().toString());
+            if (enricher != null) {
+                enricher.accept(plan);
+            }
+            completePlans.add(plan);
+        }
+        return completePlans;
+    }
+
+    protected List<Function> prepareFunctionsStaging(AutomationPackage newPackage, AutomationPackageArchive automationPackageArchive, AutomationPackageContent packageContent, ObjectEnricher enricher, AutomationPackage oldPackage) {
+        List<Function> oldFunctions = oldPackage != null ? getPackageFunctions(oldPackage) : new ArrayList<>();
+        Map<String, ObjectId> functionNameToIdMap = createNameToIdMap(oldFunctions);
+
+        List<Function> completeFunctions = new ArrayList<>();
+        for (AutomationPackageKeyword keyword : packageContent.getKeywords()) {
```

Comment:
@jeromecomte 
It is so, because in `AutomationPackageKeyword` we return the parsed keyword (`Function`) + additional map with properites, which cannot be applied immediately during deserialization (for example, resources uploaded via `ResourceManager` in `AutomationPackageManager#prepareFunctionsStaging`).

For Plans previously this wasn't required, because the `YamlPlanReader` was capable to deserialize the complete `Plan` object. But now we also have to support linked resoures in `ForEach` and `DataSet` operators, so probably we should also return additional attributes along with plans.

Reply:
@iegorov777 Get it. In this case, yes, I would align both implementations